File: dotnet/ConsoleApplication1/SaltEdge.cs
using System;
using System.Net;
using System.IO;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.OpenSsl;
using Org.BouncyCastle.Security;
using Newtonsoft.Json;

namespace SESample
{
    class Callback
    {
        private RsaKeyParameters publicKey;

        public Callback(String publicKeyPath) {
            publicKey = readPublicKey(publicKeyPath);
        }

        public bool verify(String url, String requestBody, String signature) {
            ISigner sig = SignerUtilities.GetSigner("SHA256withRSA");
            sig.Init(false, publicKey);

            String input = string.Format("{0}|{1}", url, requestBody);
            byte[] bytes = System.Text.Encoding.UTF8.GetBytes(input);

            byte[] decodedSignature = Convert.FromBase64String(signature);

            sig.BlockUpdate(bytes, 0, bytes.Length);
            return sig.VerifySignature(decodedSignature);
        }

        private RsaKeyParameters readPublicKey(string fileName)
        {
            RsaKeyParameters keyPair;

            using (var reader = File.OpenText(fileName))
                keyPair = (RsaKeyParameters)new PemReader(reader).ReadObject();

            return keyPair;
        }
    }

    class SaltEdge
    {
        const int REQUEST_EXPIRES_MINUTES = 3;
        const string APP_ID           = "APP_ID";
        const string SECRET           = "SECRET";
        const string PRIVATE_KEY_PATH = @"c:\keys\private.pem";

        private static AsymmetricKeyParameter PRIVATE_KEY = null;

        static SaltEdge()
        {
            PRIVATE_KEY = readPrivateKey(PRIVATE_KEY_PATH);
        }

        public static string get(String url) {
            WebRequest request              = buildRequest("GET", url);
            int expires                     = generateExpiresAt();
            request.Headers["Signature"]    = generateSignature("GET", expires, url, "");
            request.Headers["Expires-at"]   = expires.ToString();

            return execute(request);
        }

        public static string post(String url, object payload)
        {
            WebRequest request = buildRequest("POST", url);
            int expires = generateExpiresAt();

            request.Headers["Expires-at"] = expires.ToString();

            using (var streamWriter = new StreamWriter(request.GetRequestStream()))
            {
                string json = JsonConvert.SerializeObject(payload);
                streamWriter.Write(json);
                request.Headers["Signature"] = generateSignature("POST", expires, url, json);
            }

            return execute(request);
        }

        private static WebRequest buildRequest(string method, string url)
        {
            WebRequest request = WebRequest.Create(url);

            request.Method = method;
            request.Headers["App-id"] = APP_ID;
            request.Headers["Secret"] = SECRET;
            request.ContentType = "application/json";

            return request;
        }

        private static int generateExpiresAt()
        {
            DateTime unixBegin = new DateTime(1970, 1, 1);
            return (Int32)DateTime.UtcNow.AddMinutes(REQUEST_EXPIRES_MINUTES).Subtract(unixBegin).TotalSeconds;
        }

        private static string execute(WebRequest request)
        {
            WebResponse response;

            try
            {
                response = request.GetResponse();
            }
            catch (WebException e)
            {
                Console.WriteLine("{0}\n{1}", e.Response, e.Response.ToString());
                response = e.Response;
            }

            return process(response);
        }

        private static string process(WebResponse response) {
            using (System.IO.Stream s = response.GetResponseStream())
            {
                using (System.IO.StreamReader sr = new System.IO.StreamReader(s))
                {
                    return sr.ReadToEnd();
                }
            }
        }

        private static string generateSignature(string method, int expires, string url, string postBody="")
        {
            string signature    = string.Format("{0}|{1}|{2}|{3}", expires, method, url, postBody);
            byte[] bytes        = System.Text.Encoding.UTF8.GetBytes(signature);
            byte[] shaSignature = sign(bytes);

            return Convert.ToBase64String(shaSignature);
        }

        private static byte[] sign(byte[] bytes)
        {
            ISigner sig = SignerUtilities.GetSigner("SHA256withRSA");
            sig.Init(true, PRIVATE_KEY);
            sig.BlockUpdate(bytes, 0, bytes.Length);
            return sig.GenerateSignature();
        }

        private static AsymmetricKeyParameter readPrivateKey(string privateKeyFileName)
        {
            AsymmetricCipherKeyPair keyPair;

            using (var reader = File.OpenText(privateKeyFileName))
                keyPair = (AsymmetricCipherKeyPair)new PemReader(reader).ReadObject();

            return keyPair.Private;
        }
    }
}

File: java/java-saltedge-example/src/com/saltedge/SaltEdge.java
package com.saltedge;

import com.google.gson.Gson;
import org.bouncycastle.openssl.PEMException;
import org.bouncycastle.openssl.PEMKeyPair;
import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter;
import org.bouncycastle.util.encoders.Base64;
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.ProtocolException;
import java.net.URL;
import java.security.*;
import java.util.Calendar;

public class SaltEdge {

	public final static int REQUEST_EXPIRES_MINUTES = 3;
	public final static String APP_ID = "APP_ID";
	public final static String SECRET = "SECRET";
	public final String PRIVATE_KEY_PATH = "private.pem";
	private static PEMKeyPair PRIVATE_KEY = null;

	public SaltEdge() {
    PRIVATE_KEY = SignHelper.readPrivateKey(PRIVATE_KEY_PATH);
  }

	// HTTP GET request
	public String get(String url) {
		HttpURLConnection con = buildRequest("GET", url);

		if (con == null) {
			return "";
		}

		// Add request header
		long expires = generateExpiresAt();
		con.setRequestProperty("Signature", generateSignature("GET", expires, url, ""));
		con.setRequestProperty("Expires-at", String.valueOf(expires));
		con.setRequestProperty("Content-Type", "application/json");
		con.setRequestProperty("Accept", "application/json");

		return processResponse(con);
	}

	// HTTP POST request
	public String post(String url, Object payload) {
		HttpURLConnection con = buildRequest("POST", url);

		if (con == null) {
			return "";
		}

		// Add request header
		long expires = generateExpiresAt();
		con.setRequestProperty("Expires-at", String.valueOf(expires));

		// Generate json
		Gson gson = new Gson();
		String json = gson.toJson(payload);

		con.setRequestProperty("Signature", generateSignature("POST", expires, url, json));
		con.setRequestProperty("Content-Type", "application/json");
		con.setRequestProperty("Accept", "application/json");
		con.setDoOutput(true);
		DataOutputStream wr;

		try {
			wr = new DataOutputStream(con.getOutputStream());
			wr.writeBytes(json);
			wr.flush();
			wr.close();
		} catch (IOException e) {
			System.out.println("IOException : " + e);
			e.printStackTrace();
		}

		return processResponse(con);
	}

	private String processResponse(HttpURLConnection con) {
		BufferedReader in;
		StringBuffer response = new StringBuffer();

		try {
			in = new BufferedReader(new InputStreamReader(con.getInputStream()));
			String inputLine;

			while ((inputLine = in.readLine()) != null) {
				response.append(inputLine);
			}
			in.close();
		} catch (IOException e) {
			System.out.println("IOException : " + e);
			e.printStackTrace();
		}

		return response.toString();
	}

	private HttpURLConnection buildRequest(String method, String url) {
		HttpURLConnection con = null;

		try {
			URL obj = new URL(url);
			con = (HttpURLConnection) obj.openConnection();
			con.setRequestMethod(method);
			con.setRequestProperty("App-id", APP_ID);
			con.setRequestProperty("Secret", SECRET);
			con.setRequestProperty("Content-Type", "application/json");
		} catch (ProtocolException e) {
			System.out.println("ProtocolException : " + e);
			e.printStackTrace();
		} catch (IOException e) {
			System.out.println("IOException : " + e);
			e.printStackTrace();
		}

    return con;
  }

	private long generateExpiresAt() {
		Calendar calendar = Calendar.getInstance();
		calendar.add(Calendar.MINUTE, REQUEST_EXPIRES_MINUTES);

		return calendar.getTimeInMillis() / 1000;
  }

	private String generateSignature(String method, long expires, String url, String postBody) {
		String signature = String.format("%d|%s|%s|%s", expires, method, url, postBody);
		byte[] bytes = signature.getBytes();
		byte[] shaSignature = null;

		try {
			shaSignature = sign(bytes);
		} catch (SignatureException e) {
			System.out.println("SignatureException : " + e);
			e.printStackTrace();
		} catch (PEMException e) {
			System.out.println("PEMException : " + e);
			e.printStackTrace();
		}

    return Base64.toBase64String(shaSignature);
  }

    private byte[] sign(byte[] bytes) throws SignatureException, PEMException {
    	KeyPair keyPair=null;
    	Signature signature = null;

		try {
			signature = Signature.getInstance("SHA256withRSA");
			signature.initSign(new JcaPEMKeyConverter().getPrivateKey(PRIVATE_KEY.getPrivateKeyInfo()));
			signature.update(bytes);
		} catch (NoSuchAlgorithmException e) {
			System.out.println("NoSuchAlgorithmException : " + e);
			e.printStackTrace();
		} catch (InvalidKeyException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (SignatureException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

  	return signature.sign();
  }
}
